crates/jstz_engine/src/script.rs

```diff
@@ -0,0 +1,137 @@
+//! This module provides an interface for JavaScript scripts in SpiderMonkey.
+//! A `Script` encapsulates the parsed and compiled form of a JavaScript program,
+//! enabling efficient execution and interaction with the JavaScript engine.
+//!
+//! ## Overview
+//! - **Source Representation**: Links to the source text, including metadata
+//!     like line numbers and column offsets for debugging / stack traces.
+//! - **Bytecode**: Contains the compiled bytecode generated during parsing,
+//!     optimized for SpiderMonkey's interpreter.
+//!
+//! For more details, refer to the [ECMAScript Specification on Scripts and Modules](https://tc39.es/ecma262/#sec-scripts).
+
+use std::{marker::PhantomData, path::Path, ptr::NonNull};
+
+use mozjs::{
+    jsapi::{Compile1, JSScript, JS_ExecuteScript},
+    jsval::{JSVal, UndefinedValue},
+    rooted,
+    rust::CompileOptionsWrapper,
+};
+
+use crate::{
+    compartment::Compartment,
+    context::{CanAlloc, Context, InCompartment},
+    AsRawPtr,
+};
+
+pub struct Script<'a, C: Compartment> {
+    script: NonNull<JSScript>,
+    marker: PhantomData<(&'a (), C)>,
+}
+
+impl<'a, C: Compartment> Script<'a, C> {
+    /// Compiles a script with a given filename and returns the compiled script.
+    /// Returns `None` if the script could not be compiled.
+
+    // TODO(https://linear.app/tezos/issue/JSTZ-210):
+    // Add support for error handling / exceptions instead of using `Option`
+    pub fn compile<S>(path: &Path, script: &str, cx: &'a mut Context<S>) -> Option<Self>
+    where
+        S: InCompartment<C> + CanAlloc,
+    {
+        let mut source = mozjs::rust::transform_str_to_source_text(script);
+        let options = unsafe {
+            CompileOptionsWrapper::new(cx.as_raw_ptr(), path.to_str().unwrap(), 1)
+        };
+
+        let script = unsafe { Compile1(cx.as_raw_ptr(), options.ptr, &mut source) };
+
+        Some(Self {
+            script: NonNull::new(script)?,
+            marker: PhantomData,
+        })
+    }
+
+    /// Evaluates a script and returns its return value
+
+    // TODO(https://linear.app/tezos/issue/JSTZ-210):
+    // Add support for error handling / exceptions instead of using `Option`
+    // TODO(https://linear.app/tezos/issue/JSTZ-211):
+    // TODO: `JSVal` is not safe, we should return a safe wrapper instead
+    pub fn evaluate<'b, S>(&self, cx: &'b mut Context<S>) -> Option<JSVal>
+    where
+        S: InCompartment<C> + CanAlloc,
+        'a: 'b,
```

Comment:
Unused 'a?

Reply:
It's used in the type of script `Script<'a, C>`